sql/sql_regex.h

```diff
@@ -0,0 +1,132 @@
+/* Copyright (c) 2016, Percona and/or its affiliates. All rights reserved.
+
+   This program is free software; you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation; version 2 of the License.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program; if not, write to the Free Software
+   Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301  USA */
+
+#ifndef _regex_list_handler_h_
+#define _regex_list_handler_h_
+
+#include <string>
+#include <algorithm>
+
+#include "my_global.h"
+#include "my_pthread.h"
+#include "my_regex.h"
+
+/*
+  Wrapper class aroung my_regex that manages concurrent acess and the lifetime
+  of the compiled regex.
+*/
+class Regex
+{
+private:
+  std::string m_pattern;
+  my_regex_t m_expr;
+  bool m_compiled;
+
+  mutable mysql_rwlock_t m_rwlock;
+
+  Regex(const Regex& other) {}
```

Comment:
Argh, of course. I believe in C++03 the canonical way is to declare private (already done) and then also omit { } so that a linker error on use would result